packages/flutter/lib/src/material/spell_check_suggestions_toolbar_layout_delegate.dart

```diff
@@ -0,0 +1,51 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart' show TextSelectionToolbarLayoutDelegate;
+
+/// Positions the toolbar below [anchor] or adjusts it higher to fit above
+/// the bottom view insets, if applicable.
+///
+/// See also:
+///
+///   * [MaterialSpellCheckSuggestionsToolbar], which uses this to position itself.
+class SpellCheckSuggestionsToolbarLayoutDelegate extends SingleChildLayoutDelegate {
+  /// Creates an instance of [SpellCheckSuggestionsToolbarLayoutDelegate].
+  SpellCheckSuggestionsToolbarLayoutDelegate({
+    required this.anchor,
+    required this.heightOffset,
+  });
+
+  /// {@macro flutter.material.MaterialSpellCheckSuggestionsToolbar.anchor}
+  ///
+  /// Should be provided in local coordinates.
+  final Offset anchor;
+
+  /// The height to adjust the toolbar position by if it were to overlap with
+  /// the bottom view insets without adjustment.
+  final double heightOffset;
+
+  @override
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
+    return constraints.loosen();
+  }
+
+  @override
+  Offset getPositionForChild(Size size, Size childSize) {
+    return Offset(
+      TextSelectionToolbarLayoutDelegate.centerOn(
+        anchor.dx,
+        childSize.width,
+        size.width,
+      ),
+      anchor.dy + heightOffset,
```

Comment:
Sort of...`heightOffset` is dependent on whether or not the toolbar can fit below/above the text field. So, this layout will have to either depend on this offset or the `BuildContext` of the toolbar in order to calculate the offset. Is there a preference there?

Reply:
I was thinking that it should be as simple as this:

```dart
      anchor.dy + childSize.height > size.height
          ? size.height - childSize.height
          : anchor.dy,
```

But I tried it and it looks like the soft keyboard's viewInsets.bottom is not included in the `size` because there is no Scaffold in this Overlay subtree.  Including that bottom inset in the Padding widget in MaterialSpellCheckSuggestionsToolbar seems to make it work.

What do you think of that approach?  My main thought here is to use the built-in layout and constraints mechanisms that Flutter provides rather than relying on static values like _kDefaultToolbarHeight and our own math.  Are you able to get rid of a bunch of that math in MaterialSpellCheckSuggestionsToolbar.build if you follow this approach?